File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/promotion/PromotionRepository.java
package com.heaton.funnyvote.data.promotion;

import com.heaton.funnyvote.database.Promotion;
import com.heaton.funnyvote.database.User;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;

public class PromotionRepository implements PromotionDataSource {

    private static final String TAG = PromotionRepository.class.getSimpleName();
    private static PromotionRepository INSTANCE = null;
    private PromotionDataSource remotePromotionSource, localPromotionSource;

    public static PromotionRepository getInstance(PromotionDataSource remotePromotionSource
            , PromotionDataSource localPromotionSource) {
        if (INSTANCE == null) {
            INSTANCE = new PromotionRepository(remotePromotionSource, localPromotionSource);
        }
        return INSTANCE;
    }

    public PromotionRepository(PromotionDataSource remotePromotionSource
            , PromotionDataSource localPromotionSource) {
        this.localPromotionSource = localPromotionSource;
        this.remotePromotionSource = remotePromotionSource;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }

    @Override
    public Observable<List<Promotion>> getPromotionList(User user) {
        Observable<List<Promotion>> localVote = localPromotionSource
                .getPromotionList(user);
        Observable<List<Promotion>> remoteVote = remotePromotionSource
                .getPromotionList(user)
                .doOnNext(promotionList -> localPromotionSource.savePromotionList(promotionList))
                .onErrorResumeNext(localVote);

        return Observable.concat(remoteVote, localVote)
                .filter(list -> list != null && !list.isEmpty())
                .firstOrError()
                .toObservable()
                .onErrorResumeNext(Observable.just(new ArrayList<>()));
    }

    @Override
    public void savePromotionList(List<Promotion> promotionList) {
        localPromotionSource.savePromotionList(promotionList);
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/promotion/PromotionRepository.kt
package com.heaton.funnyvote.data.promotion

import com.heaton.funnyvote.database.Promotion
import com.heaton.funnyvote.database.User

class PromotionRepository(
        private val remotePromotionSource: PromotionDataSource,
        private val localPromotionSource: PromotionDataSource
) : PromotionDataSource {

    override fun getPromotionList(user: User, callback: PromotionDataSource.GetPromotionsCallback) {
        remotePromotionSource.getPromotionList(user, object : PromotionDataSource.GetPromotionsCallback {
            override fun onPromotionsLoaded(promotionList: List<Promotion>) {
                callback.onPromotionsLoaded(promotionList)
                localPromotionSource.savePromotionList(promotionList)
            }

            override fun onPromotionsNotAvailable() {
                localPromotionSource.getPromotionList(user, object : PromotionDataSource.GetPromotionsCallback {
                    override fun onPromotionsLoaded(promotionList: List<Promotion>) {
                        callback.onPromotionsLoaded(promotionList)
                    }

                    override fun onPromotionsNotAvailable() {
                        callback.onPromotionsNotAvailable()
                    }
                })
            }
        })
    }

    override fun savePromotionList(promotionList: List<Promotion>) {
        localPromotionSource.savePromotionList(promotionList)
    }

    companion object {

        @JvmField
        val TAG: String? = PromotionRepository::class.java.simpleName
        private var INSTANCE: PromotionRepository? = null

        @JvmStatic
        fun getInstance(remotePromotionSource: PromotionDataSource
                        , localPromotionSource: PromotionDataSource): PromotionRepository {
            return INSTANCE
                    ?: PromotionRepository(remotePromotionSource, localPromotionSource)
                            .apply {
                                INSTANCE = this
                            }
        }

        fun destroyInstance() {
            INSTANCE = null
        }

    }
}
